File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool dfs(vector<int>&a, int i, int sm, vector<vector<int>>&dp)
    {
        if (sm == 0) return 1;
        if (sm < 0 or i < 0) return 0;
        if (dp[i][sm] != -1) return dp[i][sm];
        int ntake = dfs(a, i-1, sm, dp);
        int take = dfs(a, i-1, sm-a[i], dp);
        return dp[i][sm] = take || ntake;
    }
    bool canPartition(vector<int>& nums) 
    {
        int sm = 0, n = nums.size();
        for (auto num : nums) sm +=  num;
        if (sm % 2) return false;
        else 
        {
            vector<vector<int>> dp(n+1, vector<int>(sm/2+1, -1));
            bool f = dfs(nums, n-1, sm/2, dp);
            return f;
        }
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cs
public class Solution {
    private bool Dfs(int[] a, int i, int sm, int[,] dp)
    {
        if (sm == 0) return true;
        if (sm < 0 || i < 0) return false;
        if (dp[i,sm] != -1) return dp[i,sm] == 1;
        bool take = Dfs(a, i-1, sm-a[i], dp);
        bool ntake = Dfs(a, i-1, sm, dp);
        dp[i,sm] = take || ntake ? 1 : 0;
        return dp[i,sm] == 1;
    }
    public bool CanPartition(int[] a) 
    {
        int n = a.Length, sm = 0;
        foreach (int item in a) 
            sm += item;
        if (sm % 2 != 0) return false;
        sm /= 2;
        int[,] dp = new int[n+1, sm+1];
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= sm; j++) 
                dp[i,j] = -1;
        }
        return Dfs(a, n-1, sm, dp);
    }
}